"inconsistent list element types" with security group ingress rule

Version: v0.12-rc1

The log is attached: 
[crash.log](https://github.com/hashicorp/terraform/files/3180957/crash.log)

Hi @demisx, thanks for reporting the issue.

There seems to be an issue wit the literal list not correctly being converted to a set internally.

The error and stack trace for reference:

```
panic: inconsistent list element types 
    cty.Object(map[string]cty.Type{"cidr_blocks":cty.List(cty.String), "description":cty.String, "from_port":cty.Number, "ipv6_cidr_blocks":cty.List(cty.String), "prefix_list_ids":cty.List(cty.String), "protocol":cty.String, "security_groups":cty.List(cty.String), "self":cty.Bool, "to_port":cty.Number}) then 
    cty.Object(map[string]cty.Type{"cidr_blocks":cty.List(cty.String), "description":cty.String, "from_port":cty.Number, "ipv6_cidr_blocks":cty.List(cty.String), "prefix_list_ids":cty.List(cty.String), "protocol":cty.String, "security_groups":cty.Set(cty.String), "self":cty.Bool, "to_port":cty.Number})

goroutine 66 [running]:
github.com/zclconf/go-cty/cty.ListVal(0xc0000cbda0, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0)
	/Users/jbardin/go/pkg/mod/github.com/zclconf/go-cty@v0.0.0-20190430221426-d36a6f0dbffd/cty/value_init.go:166 +0x4bf
github.com/hashicorp/terraform/lang/blocktoattr.(*fixupBlocksExpr).Value(0xc0008c6cc0, 0xc0001e7ea0, 0xc0005ad1e8, 0x7, 0xc0008c8170, 0xc0000dc301, 0x2bebaa0, 0xc0000dc380, 0x2456f20)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/lang/blocktoattr/fixup.go:166 +0x68f
github.com/hashicorp/hcl2/hcldec.(*AttrSpec).decode(0xc00084d140, 0xc0000cbb60, 0x0, 0x0, 0x0, 0xc0001e7ea0, 0x0, 0x0, 0x2456f20, 0xc0009babc0, ...)
	/Users/jbardin/go/pkg/mod/github.com/hashicorp/hcl2@v0.0.0-20190514214226-6a61d80ae3d0/hcldec/spec.go:196 +0xa5
github.com/hashicorp/hcl2/hcldec.ObjectSpec.decode(0xc00084cf90, 0xc0000cbb60, 0x0, 0x0, 0x0, 0xc0001e7ea0, 0xc0003d5428, 0xc0008c6270, 0x20, 0x20, ...)
	/Users/jbardin/go/pkg/mod/github.com/hashicorp/hcl2@v0.0.0-20190514214226-6a61d80ae3d0/hcldec/spec.go:78 +0x239
github.com/hashicorp/hcl2/hcldec.decode(0x2beae20, 0xc0001e7ec0, 0x0, 0x0, 0x0, 0xc0001e7ea0, 0x2beb820, 0xc00084cf90, 0x0, 0x0, ...)
	/Users/jbardin/go/pkg/mod/github.com/hashicorp/hcl2@v0.0.0-20190514214226-6a61d80ae3d0/hcldec/decode.go:21 +0x118
github.com/hashicorp/hcl2/hcldec.Decode(...)
	/Users/jbardin/go/pkg/mod/github.com/hashicorp/hcl2@v0.0.0-20190514214226-6a61d80ae3d0/hcldec/public.go:15
github.com/hashicorp/terraform/lang.(*Scope).EvalBlock(0xc0001e4e10, 0x2bea6a0, 0xc00084cf60, 0xc0009ba4b0, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc000070690, ...)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/lang/eval.go:68 +0x2aa
github.com/hashicorp/terraform/terraform.(*BuiltinEvalContext).EvaluateBlock(0xc00092ed00, 0x2bea6a0, 0xc00084cf60, 0xc0009ba4b0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/terraform/eval_context_builtin.go:277 +0x2b6
github.com/hashicorp/terraform/terraform.(*EvalValidateResource).Eval(0xc0001e4dc0, 0x2c18580, 0xc00092ed00, 0x2, 0x2, 0x0, 0x0)
	/Users/jbardin/go/src/github.com/hashicorp/terraform/terraform/eval_validate.go:413 +0x5a7
```

--- a/builtin/providers/test/resource_config_mode.go
+++ b/builtin/providers/test/resource_config_mode.go
@@ -28,6 +28,24 @@ func testResourceConfigMode() *schema.Resource {
 					},
 				},
 			},
+			"nested_set": {
+				Type:       schema.TypeSet,
+				Optional:   true,
+				ConfigMode: schema.SchemaConfigModeAttr,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"value": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"set": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+			},
 		},
 	}
 }

--- a/builtin/providers/test/resource_config_mode_test.go
+++ b/builtin/providers/test/resource_config_mode_test.go
@@ -93,3 +93,28 @@ resource "test_resource_config_mode" "foo" {
 		},
 	})
 }
+
+func TestResourceConfigMode_nestedSet(t *testing.T) {
+	resource.UnitTest(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckResourceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: strings.TrimSpace(`
+resource "test_resource_config_mode" "foo" {
+	resource_as_attr = []
+
+	nested_set {
+		value = "a"
+	}
+	nested_set {
+		value = "b"
+		set = []
+	}
+}
+				`),
+				Check: resource.ComposeTestCheckFunc(),
+			},
+		},
+	})
+}

--- a/go.mod
+++ b/go.mod
@@ -107,7 +107,7 @@ require (
 	github.com/xanzy/ssh-agent v0.2.1
 	github.com/xiang90/probing v0.0.0-20160813154853-07dd2e8dfe18 // indirect
 	github.com/xlab/treeprint v0.0.0-20161029104018-1d6e34225557
-	github.com/zclconf/go-cty v0.0.0-20190430221426-d36a6f0dbffd
+	github.com/zclconf/go-cty v0.0.0-20190516203816-4fecf87372ec
 	go.uber.org/atomic v1.3.2 // indirect
 	go.uber.org/multierr v1.1.0 // indirect
 	go.uber.org/zap v1.9.1 // indirect

